### PR TITLE
[Iissue-78] Add NetworkListener on virtual case

### PR DESCRIPTION
**Pull request by marifunf**
Signed-off-by: funfrockmar <marianne.funfrock@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ProjectCase.java [hunk 1]
```diff
@@ -19,6 +20,8 @@
 
     Network getNetwork();
 
+    Network getNetwork(NetworkListener listener);
```
[the remark this answers is not in the document]
Comment (sylvlecl) [reply afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ProjectCase.java:23]: We must also mention in the javadoc that the contract of being notified may not be honored by all implementations (see remote service cache). The user must check with the cache implementation he will use.

File: afs-network/afs-network-client/src/main/java/com/powsybl/afs/network/client/RemoteNetworkCacheService.java [hunk 2]
```diff
@@ -65,6 +66,11 @@ private RemoteServiceConfig getConfig() {
         return Objects.requireNonNull(configSupplier.get()).orElseThrow(() -> new AfsException("Remote service config is missing"));
     }
 
+    @Override
+    public <T extends ProjectFile & ProjectCase> Network getNetwork(T projectCase, NetworkListener listener) {
+        return getNetwork(projectCase);
```
Comment (sylvlecl) [inline afs-network/afs-network-client/src/main/java/com/powsybl/afs/network/client/RemoteNetworkCacheService.java:71]: We must at least mention in javadoc that listeners will not be added to the network.

File: afs-ext-base/src/main/java/com/powsybl/afs/ext/base/VirtualCase.java [hunk 3]
```diff
@@ -69,6 +70,11 @@ public Network getNetwork() {
         return findService(NetworkCacheService.class).getNetwork(this);
     }
 
+    @Override
+    public Network getNetwork(NetworkListener listener) {
```
Comment (sylvlecl) [inline afs-ext-base/src/main/java/com/powsybl/afs/ext/base/VirtualCase.java:74]: We must add some unit tests : 
maybe add a unit test for virtual cases to check that the listeners are indeed notified of modifications applied by the script (is it your use case?).

File: afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ImportedCase.java [hunk 4]
```diff
@@ -79,6 +80,11 @@ public Network getNetwork() {
         return findService(NetworkCacheService.class).getNetwork(this);
     }
 
+    @Override
+    public Network getNetwork(NetworkListener listener) {
```
Comment (sylvlecl) [inline afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ImportedCase.java:84]: We must add some unit test : maybe just check that listeners have indeed been registered with the returned network ?

**Review by sylvlecl — CHANGES_REQUESTED** [review]
Some tests and documentation on the behaviour need to be added

File: afs-ext-base/src/main/java/com/powsybl/afs/ext/base/VirtualCase.java (hunk 3 above)
Comment (marifunf) [reply afs-ext-base/src/main/java/com/powsybl/afs/ext/base/VirtualCase.java]: ok

File: afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ImportedCase.java (hunk 4 above)
Comment (marifunf) [reply afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ImportedCase.java]: ok

File: afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ProjectCase.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (marifunf) [reply afs-ext-base/src/main/java/com/powsybl/afs/ext/base/ProjectCase.java]: ok for the list
Is there a javadoc for powsybl-afs ?

File: afs-network/afs-network-client/src/main/java/com/powsybl/afs/network/client/RemoteNetworkCacheService.java (hunk 2 above)
Comment (marifunf) [reply afs-network/afs-network-client/src/main/java/com/powsybl/afs/network/client/RemoteNetworkCacheService.java:72]: Is there a javadoc for powsybl-afs ?

**Review by berthaultval — APPROVED** [review]

**Review by sylvlecl — APPROVED** [review]

**Pull request merged**